### ISSUE TITLE
Using diffevo for black-box optimisation.

### ISSUE THREAD

**Issue by norsedrunkensailor** [Zhangyanbo/diffusion-evolution#2]
Hellooo, 

Love this work and would love to apply it on some other data where there is not a closed form expression for the fitness function but I can evaluate a black-box function. How would you perform just a single loop to then gather more data for the next iteration? Somehow linking a continuous latent space x (learned representation) with a fitness y.  

Is this even possible with your method? 

All the best,
NDS

**Comment by Zhangyanbo** [comment]
Hi! I'm not sure if I fully understand your question. Are you asking further training after generating the final sample (i.e., $x_0$)? For this question, there might be few options. One option is to simply increase the for-loop length. Another option is you can re-diffuse (see "[Restart Sampling for Improving Generative Processes](https://proceedings.neurips.cc/paper_files/paper/2023/hash/f2543511e5f4d4764857f9ad833a977d-Abstract-Conference.html)"), which add some amount of noise to get $x_\tau$, then denoise from $x_\tau$ instead of $x_T$.

Can you explain more about "Somehow linking a continuous latent space x (learned representation) with a fitness y."? I didn't get your question.

**Comment by norsedrunkensailor** [comment]
hii, so perhaps I am not clear enough in my formulation: Your paper draws parallels between evolutionary algorithms and diffusion models and thus it implies that a diffusion model could be used to optimise some black-box fitness function. This function takes as input a latent representation x and spits out a fitness value y. In your pseudo-code you provide the key overview: For T evolution steps you first randomly initialise a population, evaluate their fitness, and then propagate the population forward. Looping over this diffusion like process your solution converges to the fitness maxima while maintaining diversity. 
Say however that I want to run just a single loop, propagate x_t and then externally evaluate the fitness of the new population (experimentally for example), is there a way to allow for this? I.e instead of something like

`x, population_trace, fitness_count = optimizer.optimize(two_peak_density, xT, trace=True)`

where we have a closed form two_peak_density, we can only evaluate the fitness function experimentally. 
So to summarise, how could I use this as an evolutionary algorithm to optimise a black-box fitness function (no closed form, slow evaluation)

Hopefully I was able to describe it better

All the best, 
Niklas

**Comment by Zhangyanbo** [comment]
Sure, you can definitely do this. Please check the `README.md` file, in the "Typical use" section, we show a way to expose the fitness evaluation:
```python
from diffevo import DDIMScheduler, BayesianGenerator
from diffevo.examples import two_peak_density

scheduler = DDIMScheduler(num_step=100)

x = torch.randn(512, 2)

for t, alpha in scheduler:
    fitness = two_peak_density(x, std=0.25)
    generator = BayesianGenerator(x, fitness, alpha)
    x = generator(noise=0)
```

So you can hack in and replace the fitness by whatever you want.

**Comment by norsedrunkensailor** [comment]
Aii, im sorry I should have seen that
Off to playing around!!!

Thank you
Best,
NDS
